master/internal/api_tensorboard.go

```diff
@@ -61,46 +88,272 @@ func (a *apiServer) KillTensorboard(
 func (a *apiServer) LaunchTensorboard(
 	ctx context.Context, req *apiv1.LaunchTensorboardRequest,
 ) (*apiv1.LaunchTensorboardResponse, error) {
-	experimentIds := make([]int, 0)
-	trialIds := make([]int, 0)
-	for _, id := range req.ExperimentIds {
-		experimentIds = append(experimentIds, int(id))
+	var err error
+
+	// Validate the request.
+	if len(req.ExperimentIds) == 0 && len(req.TrialIds) == 0 {
+		err = errors.New("must set experiment or trial ids")
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-	for _, id := range req.TrialIds {
-		trialIds = append(trialIds, int(id))
+	exps, err := getTensorBoardConfigsFromReq(a.m.db, req)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	if len(exps) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "no experiments found")
 	}
 
-	params, err := a.prepareLaunchParams(ctx, &protoCommandParams{
+	spec, err := a.getCommandLaunchParams(ctx, &protoCommandParams{
 		TemplateName: req.TemplateName,
 		Config:       req.Config,
 		Files:        req.Files,
 		MustZeroSlot: true,
 	})
 	if err != nil {
-		return nil, api.APIErr2GRPC(err)
+		return nil, api.APIErr2GRPC(errors.Wrapf(err, "failed to prepare launch params"))
 	}
 
-	tensorboardLaunchReq := command.TensorboardRequest{
-		CommandParams: params,
-		ExperimentIDs: experimentIds,
-		TrialIDs:      trialIds,
+	// Postprocess the spec.
+	spec.Config.Description = fmt.Sprintf(
```

Comment:
I don't want to split this function into different helper functions since the logic will be only used here. I think it's fine to make one function 200 lines as long as it's single-purpose.